### PR TITLE
Rewrites remote AI eye movement code , general FOB Drone QOL

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## About The Pull Request
Rewrites remote AI eye movement code so i have finer control over it (and it doesn't skip tiles anymore at high acceleration)

FOB Drone now uses bubble alerts , and is slower because its way too fast.

Draft because i still have to figure out how to make the drone face an atom the controller clicks on , as an extra qol.
Edit : figured out.

## Why It's Good For The Game
More QoL, easier life for engies.

## Changelog
:cl:
qol: Slowed down the FOB Drone, gave it bubble alerts and fixed double .remaining messages
qol: Drones now face the direction of any turf you click.
code: Rewrote the remote AI eye code to support more configurable move_delays
code: Rewrote the remote AI eye acceleration , it now gets removed on direction change.
/:cl:

### CODE REVIEW

**Comment by MereWalker** [comment]
This will make me not dread building fob as an engineer now. Thank you!

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 1]
```diff
@@ -205,9 +205,11 @@
 /mob/camera/aiEye/remote
 	name = "Inactive Camera Eye"
 	ai_detector_visible = FALSE
-	var/sprint = 10 //This number is not doing anything if it's not a multiple of 20
+	var/move_delay = 0.1 SECONDS// quite fast
+	var/tiles_moved = 0 // Tiles moved in 1 direction , used for smart acceleration
+	var/max_tile_acceleration = 8 // a limit to how far acceleration can go
```
[the remark this answers is not in the document]
Comment (MLGTASTICa) [reply code/game/objects/machinery/computer/camera_advanced.dm:213]: Added autodoc explanations , fine with you?

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 2]
```diff
@@ -263,27 +266,21 @@
...
-	cooldown = world.timeofday + 0.5 SECONDS
+	if(cooldown > world.time)
+		return
+	cooldown = world.time +	move_delay
 	if(acceleration)
-		sprint = min(sprint + 0.5, max_sprint)
-	else
-		sprint = initial
-
+		cooldown -= tiles_moved * (move_delay/10)
```
[the remark this answers is not in the document]
Comment (MLGTASTICa) [reply code/game/objects/machinery/computer/camera_advanced.dm]: done.

File: code/modules/mining/remote_fob/remote_fob_computer.dm [hunk 3]
```diff
@@ -121,6 +121,8 @@
 	if(!spawn_spot)
 		CRASH("CreateEye() called without a spawn_spot designated")
 	eyeobj = new /mob/camera/aiEye/remote/fobdrone(get_turf(spawn_spot))
+	eyeobj.move_delay = 0.2 SECONDS
+	eyeobj.acceleration = FALSE
```
[the remark this answers is not in the document]
Comment (MLGTASTICa) [reply code/modules/mining/remote_fob/remote_fob_computer.dm]: i made a define and put it in construction.dm , not really sure if its a good spot.

File: code/modules/mining/remote_fob/remote_fob_computer.dm [hunk 4]
```diff
@@ -193,3 +198,8 @@
 		user.unset_interaction()
 		return
 	return ..()
+
+// Lets players click a tile while controlling to face it.
+/obj/machinery/computer/camera_advanced/remote_fob/proc/drone_face(datum/source, atom/target, turf/location, control, params)
```
[the remark this answers is not in the document]
Comment (MLGTASTICa) [reply code/modules/mining/remote_fob/remote_fob_computer.dm]: renamed to it on_controller_click.

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 5]
```diff
@@ -263,27 +266,21 @@
...
+	var/turf/T = get_turf(get_step(src, direct))
+	// check for dir change , if we changed then remove all acceleration
+	if(get_dir(src, T) != dir)
+		tiles_moved = 0
+	if(tiles_moved < max_tile_acceleration)
+		tiles_moved++
```
[the remark this answers is not in the document]
Comment (MLGTASTICa) [reply code/game/objects/machinery/computer/camera_advanced.dm]: why? (considering the new code)

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 6]
```diff
@@ -205,9 +205,11 @@
 /mob/camera/aiEye/remote
 	name = "Inactive Camera Eye"
 	ai_detector_visible = FALSE
-	var/sprint = 10 //This number is not doing anything if it's not a multiple of 20
+	var/move_delay = 0.1 SECONDS// quite fast
+	var/tiles_moved = 0 // Tiles moved in 1 direction , used for smart acceleration
+	var/max_tile_acceleration = 8 // a limit to how far acceleration can go
 	var/cooldown = 0
-	var/acceleration = FALSE
+	var/acceleration = TRUE
```
[the remark this answers is not in the document]
Comment (MLGTASTICa) [reply code/game/objects/machinery/computer/camera_advanced.dm]: Added.

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 7]
```diff
@@ -263,27 +269,20 @@
...
-	if(acceleration)
-		sprint = min(sprint + 0.5, max_sprint)
-	else
-		sprint = initial
-
+	if(cooldown > world.time)
+		return
+	acceleration = ((cooldown + move_delay * 5) > world.time) ? acceleration : 0
+	cooldown = world.time +	move_delay
+	cooldown -= acceleration ? tiles_moved * (move_delay / 10) : 0
```
Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:279]: Why are you doing it in two lines

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 8]
```diff
@@ -263,27 +269,20 @@
...
+	var/turf/T = get_turf(get_step(src, direct))
+	// check for dir change , if we changed then remove all acceleration
+	if(get_dir(src, T) != dir)
+		tiles_moved = 0
+	tiles_moved = tiles_moved < max_tile_acceleration ? tiles_moved++ : tiles_moved
```
Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:284]: This is a min.

File: code/modules/mining/remote_fob/remote_fob_computer.dm [hunk 9]
```diff
@@ -121,6 +121,8 @@
 	if(!spawn_spot)
 		CRASH("CreateEye() called without a spawn_spot designated")
 	eyeobj = new /mob/camera/aiEye/remote/fobdrone(get_turf(spawn_spot))
+	eyeobj.move_delay = FOBDRONE_MOVEDELAY
```
Comment (BraveMole) [inline code/modules/mining/remote_fob/remote_fob_computer.dm:124]: Why the hell is this done at runtime

File: code/__DEFINES/construction.dm [hunk 10]
```diff
@@ -36,3 +36,6 @@
 
 //The amount of materials you get from a sheet of mineral like iron/diamond/glass etc
 #define MINERAL_MATERIAL_AMOUNT 2000
+
+// The movement delay of the drone used to built FOBS
+#define FOBDRONE_MOVEDELAY 0.2 SECONDS
```
Comment (BraveMole) [inline code/__DEFINES/construction.dm:41]: I said, why is it not on the /mob/camera/aiEye/remote/fobdrone, in the constructor

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/computer/camera_advanced.dm (hunk 7 above)
Comment (MLGTASTICa) [reply code/game/objects/machinery/computer/camera_advanced.dm:279]: moved it onto 1 line. It was mainly for clarity.

File: code/modules/mining/remote_fob/remote_fob_computer.dm (hunk 9 above)
Comment (MLGTASTICa) [reply code/modules/mining/remote_fob/remote_fob_computer.dm:124]: moved it to the object declaration.

File: code/__DEFINES/construction.dm (hunk 10 above)
Comment (MLGTASTICa) [reply code/__DEFINES/construction.dm:41]: removed the def and put it in the object  declaration.

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 11]
```diff
@@ -263,27 +269,19 @@
...
-	cooldown = world.timeofday + 0.5 SECONDS
-	if(acceleration)
-		sprint = min(sprint + 0.5, max_sprint)
-	else
-		sprint = initial
-
+	if(cooldown > world.time)
+		return
+	acceleration = ((cooldown + move_delay * 5) > world.time) ? acceleration : 0
+	cooldown = world.time +	move_delay - acceleration ? tiles_moved * (move_delay / 10) : 0
```
Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:278]: cooldown = world.time +	move_delay * (1 - acceleration * tiles_moved / 10)

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 12]
```diff
@@ -263,27 +269,19 @@
...
-
-	cooldown = world.timeofday + 0.5 SECONDS
-	if(acceleration)
-		sprint = min(sprint + 0.5, max_sprint)
-	else
-		sprint = initial
-
+	if(cooldown > world.time)
+		return
+	acceleration = ((cooldown + move_delay * 5) > world.time) ? acceleration : 0
```
Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:277]: acceleration = !!((cooldown + move_delay * 5) > world.time)

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/computer/camera_advanced.dm (hunk 11 above)
Comment (MLGTASTICa) [reply code/game/objects/machinery/computer/camera_advanced.dm]: done.

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 13]
```diff
@@ -205,9 +205,14 @@
 /mob/camera/aiEye/remote
 	name = "Inactive Camera Eye"
 	ai_detector_visible = FALSE
-	var/sprint = 10 //This number is not doing anything if it's not a multiple of 20
+	/// The delay applied after moving to a tile.
+	var/move_delay = 0.1 SECONDS// quite fast
+	/// Internal variable used to keep track of the amount of tiles we have moved in the same direction
+	var/tiles_moved = 0 // Tiles moved in 1 direction , used for smart acceleration
```
Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:211]: Right comment is uneeded

Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:211]: Ditto

File: code/game/objects/machinery/computer/camera_advanced.dm [hunk 14]
```diff
@@ -205,9 +205,14 @@
...
+	/// Limits tiles_moved to this value.
+	var/max_tile_acceleration = 8 // a limit to how far acceleration can go
```
Comment (BraveMole) [inline code/game/objects/machinery/computer/camera_advanced.dm:213]: Ditto

File: code/modules/mining/remote_fob/remote_fob_computer.dm [hunk 15]
```diff
@@ -193,3 +196,8 @@
 		user.unset_interaction()
 		return
 	return ..()
+
+// Lets players click a tile while controlling to face it.
```
Comment (BraveMole) [inline code/modules/mining/remote_fob/remote_fob_computer.dm:200]: Autodoc is three /

**Review by BraveMole — APPROVED** [review]

**Pull request merged**